src/ingest-pipeline/airflow/dags/error_catching/failure_callback.py

```diff
@@ -0,0 +1,171 @@
+import json
+from pprint import pprint
+from requests.exceptions import HTTPError
+from requests import codes
+import traceback
+
+from airflow.providers.http.hooks.http import HttpHook
+from airflow.utils.email import send_email
+
+from utils import get_auth_tok
+
+"""
+TODO
+- this requires crypt_auth_tok to be an op_kwarg, is that reasonable?
+- fix DRY issues (e.g. get_submission_context?, set_status)
+- type hinting
+"""
+
+
+class FailureCallbackException(Exception):
+    pass
+
+
+class FailureCallback:
+    """
+    List should be overwritten by each subclass with appropriate values.
+    """
+    external_exceptions = []
+    # TODO: Switch to curator email(s)
+    internal_email_recipients = ["gesina@psc.edu"]
+
+    def __init__(self, context, execute_methods=True):
+        self.context = context
+        self.uuid = self.context.get("task_instance").xcom_pull(key="uuid")
+        self.auth_tok = get_auth_tok(**{"crypt_auth_tok": self.context.get("crypt_auth_tok")})
+        self.dag_run = self.context.get("dag_run")
+        self.task = self.context.get("task")
+        self.exception = self.context.get("exception")
+        self.exception_name = type(self.exception).__name__
+
+        if execute_methods:
+            self.set_status()
+            self.send_notifications()
+
+    def send_notifications(self):
+        """
+        Subclasses should override the send_notifications method
+        in order to add self.send_asana_notification() or other
+        custom notifications.
+        """
+        self.send_failure_email()
+
+    # This is simplified from pythonop_get_dataset_state in utils
+    def get_submission_context(self):
+        method = "GET"
+        headers = {
+            "authorization": f"Bearer {self.auth_tok}",
+            "content-type": "application/json",
+            "X-Hubmap-Application": "ingest-pipeline",
+        }
+        http_hook = HttpHook(method, http_conn_id="ingest_api_connection")
```

Comment:
Maybe using entity_api_connection would be better, since we are moving toward the standardized endpoints suggested by Pitt

Reply:
Done!